src/mscorlib/src/System/Text/Encoding.cs

```diff
@@ -88,14 +88,54 @@ namespace System.Text
 #endif
     public abstract class Encoding : ICloneable
     {
-        private static volatile Encoding defaultEncoding;
-        private static volatile Encoding unicodeEncoding;
-        private static volatile Encoding bigEndianUnicode;
-        private static volatile Encoding utf7Encoding;
-        private static volatile Encoding utf8Encoding;
-        private static volatile Encoding utf32Encoding;
-        private static volatile Encoding asciiEncoding;
-        private static volatile Encoding latin1Encoding;
+        // We lazily-initialize the static Encoding properties by storing them in
+        // nested static classes. The static constructor, which holds the initialization
+        // logic, will only be run on the first access to a static member of the class
+        // (see dotnet/coreclr#1193 for more info). Thus, we get the benefits of both
+        // thread-safe and lock-free initialization.
+
+        // Please do not aggregate these classes into one static class. Otherwise,
+        // all of these encodings will be initialized when only one of them is needed.
+
+        private static class ASCIIHolder
+        {
```

Comment:
It is wasteful to create new holder types for these - even trivial type costs 100's bytes at runtime. It would better to make these readonly variables in the respective encoding types:

```
public class ASCIIEncoding
{
    internal readonly ASCIIEncoding Instance = new ASCIIEncoding();
...
```

Similar pattern is used in number of places in coreclr and corefx.